includes/content.php

```diff
@@ -179,96 +191,139 @@ function pmpro_search_filter($query)
-					$my_pages = array_unique(array_merge($my_pages, $member_pages));
-				}
-            } // foreach
-        } // if($levels)
+	// If posts aren't being filtered, don't filter categories either.
+	if ( ! in_array( 'post', $pmpro_search_filter_post_types ) && $query->is_category() ) {
+		return $query;
+	}
+	
+	// If posts aren't being filtered, don't filter tags either.
+	if ( ! in_array( 'post', $pmpro_search_filter_post_types ) && $query->is_tag() ) {
+		return $query;
+	}
 
-        //get hidden page ids
-        if( ! empty( $my_pages ) ) {
-			$sql = "SELECT page_id FROM $wpdb->pmpro_memberships_pages WHERE page_id NOT IN(" . implode(',', array_map( 'esc_sql', $my_pages ) ) . ")";
-		} else {
-			$sql = "SELECT page_id FROM $wpdb->pmpro_memberships_pages";
-		}
-        $hidden_page_ids = array_values(array_unique($wpdb->get_col($sql)));
-		
-        if( $hidden_page_ids ) {
-			//avoiding post_parent queries for now
-			if( empty( $query->query_vars['post_parent'] ) ) {
-				$query->set( 'post__not_in', array_merge( $query->get('post__not_in'), $hidden_page_ids ) );
-			}
-		}
-				
-        //get categories that are filtered by level, but not my level
-        global $pmpro_my_cats;
-		$pmpro_my_cats = array();
-
-        if( $levels ) {
-            foreach( $levels as $key => $level ) {
-                $member_cats = pmpro_getMembershipCategories($level->id);
-                $pmpro_my_cats = array_unique(array_merge($pmpro_my_cats, $member_cats));
-            }
-        }
-		
-        //get hidden cats
-        if( ! empty( $pmpro_my_cats ) ) {
-			$sql = "SELECT category_id FROM $wpdb->pmpro_memberships_categories WHERE category_id NOT IN(" . implode(',', array_map( 'esc_sql', $pmpro_my_cats ) ) . ")";
-		} else {
-			$sql = "SELECT category_id FROM $wpdb->pmpro_memberships_categories";
-		}							
-        $hidden_cat_ids = array_values(array_unique($wpdb->get_col($sql)));
-		
-        //make this work
-        if( $hidden_cat_ids ) {
-
-			// Get all registered category ID's so we may remove the ones we need to hide. This is to support posts that may belong to multiple categories and only one of them is hidden.
-			$all_cat_ids = get_terms( array(
-				'taxonomy' => 'category',
-				'fields'   => 'ids',
-				'get'      => 'all',
-			));
-
-			// Show only these categories to members.
-			$query->set( 'category__in', array_merge( $query->get( 'category__in' ), array_diff( $all_cat_ids, $hidden_cat_ids ) ) );
-						
-			//filter so posts in this member's categories are allowed
-			add_action( 'posts_where', 'pmpro_posts_where_unhide_cats' );
-		}
-    }
+	/**
+	 * Okay. We're going to filter the search results.
+	 * Some explaination of what we're doing.
+	 * A = All posts hidden by level.
+	 * B = All posts the current user has access to by level.
+	 * C = All posts hidden by category.
+	 * D = All posts the current user has access to by category.
+	 * 
+	 * Then hidden posts = (A - B) + (C - D)
```

Comment:
Should this be `(A + C) - (B + D)`? In other words `(All restricted posts) - (All restricted posts that the user has access to)`?

Imagine a case where a post is restricted to level `1` but also in a restricted category `restricted_cat`. If a user had level 1 but did not have access to posts in `restricted_cat`, would we want to show this post?

If we flip the situation and the user has access to `restricted_cat` but they don't have level `1`, would we want to show the post now?

Reply:
You're right. I was trying to get this done in 2 queries vs 4. I think we might have to do 4 queries... or (A - B - D) + (C - B - D) which would be 2 queries. Need to try that out and see how complicated those queries get.

Reply:
Latest commit does it in 4 separate queries now. I tried doing just 2 with the (A +C) - (B + D) approach, but this requires joining the relatively smaller pmpro tables with the much larger core WP tables in ways I suspect would make performance worse in a larger % of cases.

Note also that the latest version of the code swaps the B & C queries/arrays with what is explained here.